### PR TITLE
Extract parser functionality from ICEndpoint into ICParser

### PR DESCRIPTION
**Pull request by EightSQ**

### CODE REVIEW

**Comment by Paulpanther** [comment]
We have established a new Protocol package and moved parsing functionality over to the new ICParser, which is as it should be a stateless method bin. The functionality has not been removed from the ICEndpoint yet, so this is not ready for pull.

**Comment by timgarrels** [comment]
Is this PR still cared for?
There was no activity for a week now. Coverage decreased and Tests are failing


@Paulpanther @EightSQ

File: packages/IMAPClient-Protocol.package/ICParser.class/class/parseHeaderChunk.for..st [hunk 1]
```diff
@@ -0,0 +1,13 @@
+parsing
+parseHeaderChunk: aHeaderChunk for: anEmail
```
Comment (timgarrels) [inline packages/IMAPClient-Protocol.package/ICParser.class/class/parseHeaderChunk.for..st]: This method name does not imply that the `anEmail`-object will be modified.
This needs to be renamed.
For example `addHeaderDataTo: anEmail from: aHeaderChunk`

File: packages/IMAPClient-Core.package/ICEndPoint.class/instance/folderPathFrom..st [hunk 2]
```diff
@@ -1,11 +1,8 @@
 protocol parse
 folderPathFrom: responseLine
```
Comment (timgarrels) [inline packages/IMAPClient-Core.package/ICEndPoint.class/instance/folderPathFrom..st]: Wasn't this replaced by `ICParser folderPathFrom:`?

File: packages/IMAPClient-Protocol.package/ICParser.class/class/protocolBlank.st [hunk 3]
```diff
@@ -0,0 +1,4 @@
+constants
+protocolBlank
```
Comment (timgarrels) [inline packages/IMAPClient-Protocol.package/ICParser.class/class/protocolBlank.st]: Is this ever called?

File: packages/IMAPClient-Protocol.package/ICParser.class/class/protocolCommandStart.st [hunk 4]
```diff
@@ -0,0 +1,4 @@
+constants
+protocolCommandStart
```
Comment (timgarrels) [inline packages/IMAPClient-Protocol.package/ICParser.class/class/protocolCommandStart.st]: Is this ever called?

File: packages/IMAPClient-Protocol.package/ICParser.class/class/protocolError.st [hunk 5]
```diff
@@ -0,0 +1,4 @@
+responses
+protocolError
```
Comment (timgarrels) [inline packages/IMAPClient-Protocol.package/ICParser.class/class/protocolError.st]: Is this ever called?

File: packages/IMAPClient-Protocol.package/ICParser.class/class/protocolParameterError.st [hunk 6]
```diff
@@ -0,0 +1,4 @@
+responses
+protocolParameterError
```
Comment (timgarrels) [inline packages/IMAPClient-Protocol.package/ICParser.class/class/protocolParameterError.st]: Is this ever called?

File: packages/IMAPClient-Protocol.package/ICParser.class/class/protocolQuote.st [hunk 7]
```diff
@@ -0,0 +1,4 @@
+constants
+protocolQuote
```
Comment (timgarrels) [inline packages/IMAPClient-Protocol.package/ICParser.class/class/protocolQuote.st]: Is this ever called?

File: packages/IMAPClient-Protocol.package/ICParser.class/class/parseFetchFlagsOnly..st [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+		ifTrue: [^ Dictionary new]. 
+		
+	"iterate through lastResponse and find flags for every uid"
+ 	^ (aResponseStringArray collect: [:line |
+		| uidPart uid |
+		uidPart := (line allRegexMatches: 'UID \d*') ifEmpty: [ ^ Dictionary new].
+		uid := (uidPart first allRegexMatches: '[\d]+') ifEmpty: [ ^ Dictionary new].
+
+		Association key: uid first value: (self flags select: [:flag |
+			line includesSubString: flag])]) as: Dictionary
```
Comment (timgarrels) [inline packages/IMAPClient-Protocol.package/ICParser.class/class/parseFetchFlagsOnly..st]: `as: Dictionary` is bad style. `asDictionary` is not implemented, so maybe use `Dictionary newFrom: (Association ...)`

File: packages/IMAPClient-Protocol.package/ICParser.class/class/parseFetchFlagsOnly..st [hunk 9]
```diff
@@ -0,0 +1,14 @@
+parsing
+parseFetchFlagsOnly: aResponseStringArray
+
+	(aResponseStringArray isEmpty) 
+		ifTrue: [^ Dictionary new]. 
+		
+	"iterate through lastResponse and find flags for every uid"
+ 	^ (aResponseStringArray collect: [:line |
+		| uidPart uid |
```
Comment (timgarrels) [inline packages/IMAPClient-Protocol.package/ICParser.class/class/parseFetchFlagsOnly..st]: Local Vars should be on top of the method

File: packages/IMAPClient-Core.package/ICEndPoint.class/instance/parseFetchFlagsOnly..st [hunk 10]
```diff
@@ -1,23 +1,16 @@
 protocol parse
 parseFetchFlagsOnly: aFolder
```
Comment (timgarrels) [inline packages/IMAPClient-Core.package/ICEndPoint.class/instance/parseFetchFlagsOnly..st]: Shouldn't the ICEndpoint be free of parsing tasks?

**Review by timgarrels — DISMISSED** [review]

**Comment by timgarrels** [comment]
`ICEndpoint` still has parse relevant constants. Those are duplicated in `ICParser` (`protocolBlank` for example)

File: packages/IMAPClient-Protocol.package/ICParser.class/class/parseHeaderChunk.for..st (hunk 1 above)
Comment (EightSQ) [reply packages/IMAPClient-Protocol.package/ICParser.class/class/parseHeaderChunk.for..st]: Resolved with `populate:withHeadersFrom:`.

Comment (EightSQ) [reply packages/IMAPClient-Protocol.package/ICParser.class/class/parseHeaderChunk.for..st]: It is.

File: packages/IMAPClient-Protocol.package/ICParser.class/class/parseFetchFlagsOnly..st (hunk 9 above)
Comment (EightSQ) [reply packages/IMAPClient-Protocol.package/ICParser.class/class/parseFetchFlagsOnly..st]: It does more than parsing. We renamed the method to `parseFetchedFlagsAndPopulate:`.

**Review by martintaraz — APPROVED** [review]

**Pull request merged**